[Feature:TAGrading] peer assignments matrix options

josharsh opened the pull request:
closes #5370 
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant

### What is the current behavior?
Randomization controls require scrolling down, options for preparing peer matrix are limited. 

### What is the new behavior?
Options for preparing peer matrix are updated
* Fix Peer Matrix Control Positions
* Add controls for adding to the CSV
* Customize Buttons for All Grades All, Clear Peer Matrix Data

josharsh commented:
@bmcutler @emaicus @jchulton 
Randomization function has gotten huge with this PR, I was hoping I would open up another PR for the sole remaining feature (so that existing work does not break) on issue #5370, (To give an option to restrict grading until submission has been made) 
This PR covers:
1 - UI modification
2 - Buttons for All Grades All, Clear Pee Matrix
3 - Option to Download Peer assignments as csv
4 - Option to Restrict Randomized Peer assignments to each registration section.

shailpatels commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -63,11 +63,96 @@ public function RandomizePeers($gradeable_id) {
+            }
+            /* Duplicating the entire functionality of randomization, but for each registration section rather than whole class" */
+            $final_grading_info = [];
+            foreach ($sections as $section) {
```
What is the purpose of the above for loop? It seems like both are just iterating through the associative array
Is it possible to do something like:
```php
foreach( $sections as $key => $value ){
 $s = $value['sections_registartion_id']
```
Could you do something like ^
To avoid the additional for loop

josharsh replied:
it works simply to get students from a registration section (Individual):
$students = $this->core->getQueries()->getUsersByRegistrationSections([$section]);

shailpatels commented on site/app/templates/admin/admin_gradeable/AdminGradeablePeers.twig:
```diff
@@ -1,10 +1,37 @@
+    <div class="card col-md-6" style="margin:0px">
+        <div class="content" style="margin:0px">Enter number of peers for one student to grade and randomly assign peer grading in class. <br>
+            <hr>
+            <input type="checkbox" class="ignore" name="restrict_to_registration" id="restrict_to_registration" value="Restrict to Registration Sections">Restrict to Registration Sections</input>
```
Lets use dash's instead of underscores for CSS id's to be consistent with the rest of submitty
e.g: `restrict-to-registration`

josharsh replied:
Done!

josharsh replied:
corrected!

shailpatels commented on site/app/templates/admin/admin_gradeable/AdminGradeablePeers.twig:
```diff
@@ -173,4 +196,35 @@
\ No newline at end of file
+    function downloadPeerDataAsCsv(table_id) {
+    // Select rows from table_id
+    var rows = document.querySelectorAll('table#' + table_id + ' tr');
```
Try and avoid using `var` when possible using `const` and `let` are safer alternatives in JS, for example, `const rows`

josharsh replied:
corrected!

shailpatels commented on site/app/templates/admin/admin_gradeable/AdminGradeablePeers.twig:
```diff
@@ -173,4 +196,35 @@
+    var csv = [];
+    header=['grader','student'];
+    csv.push(header.join(','));
+    for (var i = 1; i < rows.length; i++) {
```
use `let  i = 1` instead of `var`, otherwise the variable i will actually be defined in the above scope and not be restricted to the for loop - this could cause variable pollution in other parts of the scope.

josharsh replied:
done

shailpatels commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -276,8 +313,10 @@ function ajaxCheckBuildStatus() {
-    var number_to_grade=$('#number_to_peer_grade').val();
+function setRandomGraders(gradeable_id,p_values,successCallback,errorCallback,all_grade_all) {
+    var number_to_grade=1;
+    if(all_grade_all===true){number_to_grade=10000;}
```
lets make these `if else` statements go in their own line e.g:

`if (all_grad_all === true){
   number_to_grade = 10000;
}else{
...
}

josharsh replied:
Done

shailpatels commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -286,13 +325,20 @@ function setRandomGraders(gradeable_id,p_values,successCallback,errorCallback) {
         type: "POST", 
         url: buildCourseUrl(['gradeable', gradeable_id, 'RandomizePeers']),
-        data: {csrf_token:p_values['csrf_token'],number_to_grade:number_to_grade},
+        data: {csrf_token:p_values['csrf_token'],number_to_grade:number_to_grade,restrict_to_registration:restrict_to_registration},
```
put the dictionary values on their own line so its easier to read
```json
data:{
  "csrf_token" : p_values[...],
  "nuber_to_grade" : ...,
  ...
}
```

josharsh replied:
corrected!

shailpatels commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -286,13 +325,20 @@ function setRandomGraders(gradeable_id,p_values,successCallback,errorCallback) {
-        data: {csrf_token:p_values['csrf_token'],number_to_grade:number_to_grade},
+        data: {csrf_token:p_values['csrf_token'],number_to_grade:number_to_grade,restrict_to_registration:restrict_to_registration},
         success: function(response){
+            console.log(response);
```
I think these logs should be removed - it doesn't look like they are displaying important information that should be displayed everytime this request is sent.

josharsh replied:
Removed all console logs

josharsh commented:
@jchulton @bmcutler  I am not sure on "Right now, students spill over into grading others outside of their registration section"
Here's the ss when I check "restrict to registration sections" and enter 50 (Which is more than number in each registration)
All students are assigned to grade everyone in their respective registration section. 

Am I misinterpreting something??
![image](https://user-images.githubusercontent.com/26869672/88520463-3c206980-d011-11ea-8e6a-e5f239a7e4bc.png)

josharsh commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -163,6 +163,7 @@ $(document).ready(function () {
                         $(val).val('0');
                     }
                     data[val.name] = $(val).val();
+                    let all_grade_all=false;
```
It is passed as a parameter to setRandomGraders():
~~~javascript

setRandomGraders($('#g_id').val(), data, function (response_data) {
                // Clear errors by setting new values
                for (let key in response_data) {
                    if (response_data.hasOwnProperty(key)) {
                        clearError(key, response_data[key]);
                    }
                }
                // Clear errors by just removing red background
                for (let key in data) {
                    if (data.hasOwnProperty(key)) {
                        clearError(key);
                    }
                }
                updateErrorMessage();
            }, updateGradeableErrorCallback, all_grade_all);
~~~

jchulton replied:
Isn't it out of scope of that function? The scope of the variable seems limited to this function...

josharsh replied:
Yes, I updated to use boolean directly (eliminating all_grade_all)

josharsh commented on site/app/templates/admin/admin_gradeable/AdminGradeablePeers.twig:
```diff
@@ -1,10 +1,37 @@
+    <div class="card col-md-12" style="margin:0px">
+        <div class="content"style="margin:0px">
+        <p>Other Controls: </p>
+        <div class="btn-wrapper" >
```
Corrected!

josharsh commented on site/app/templates/admin/admin_gradeable/AdminGradeablePeers.twig:
```diff
@@ -1,10 +1,37 @@
+<div id="peer-control">
+    <h1> Options for Peer Matrix </h1>
+    <div class="row">
+    <div class="card col-md-6" style="margin:0px">
```
Corrected!

josharsh commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -178,13 +179,49 @@ $(document).ready(function () {
         }
             
         });
+        $('#all_grade_all').click(
```
Corrected!

josharsh commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -63,11 +63,96 @@ public function RandomizePeers($gradeable_id) {
+            foreach ($sections as $i => $section) {
+                $sections[$i] = $section['sections_registration_id'];
+            }
+            /* Duplicating the entire functionality of randomization, but for each registration section rather than whole class" */
```
Helper function was always an option but there are many boolean(s) here
1 -  All grade all when Restricted to registration section
2 -  All grade all when Restricted to Submission before Grading
3 - All grade all when restricted to registration section an also submission before grading
4 - All grades all when no options are selected
I thought it will only make everything more complex so thought of doing everything with same set of variables inside one function (with proper comments) 
If you ask me to, I can look for a good way to implement is using a helper function.

jchulton replied:
Again, don't let it hold you up. The code still works (:

jchulton commented:
The Restrict to Registration section is working properly.

josharsh commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -178,13 +179,49 @@ $(document).ready(function () {
+                if ( confirm("Each student grades every other student! Continue?")) {
+                    let data = {'csrf_token': csrfToken};
+                    data[this.name] = $(this).val();
+                    let all_grade_all=true;
```
corrected!

josharsh commented:
@jchulton Fixed the code as per reviews
* Removed not required variables
* Indentation Fixes

josharsh commented:
@jchulton 
* Fixed Team Assignments with Submit Before Grading 
* Fixed Lesser submission edge case
* Fixed edge case with restrict to registration section and submit before grading both checked.

jchulton commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -286,12 +326,27 @@ function setRandomGraders(gradeable_id,p_values,successCallback,errorCallback) {
+            submit_before_grading:submit_before_grading,
+        },
         success: function(response){
+            console.log(response);
```
extraneous log readded

josharsh replied:
Removed!

jchulton replied:
Looks good

josharsh commented:
![image](https://user-images.githubusercontent.com/26869672/88956084-75c7cd80-d2ba-11ea-8426-0446775d4e7d.png)
Updated UI
@jchulton @bmcutler

jchulton commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -63,11 +63,110 @@ public function RandomizePeers($gradeable_id) {
+                }
+                
+                $number_of_students = count($student_array);
+                /*if ($number_of_students <= 1) {
```
Please remove commented out code

josharsh replied:
Removed!

jchulton commented on site/public/js/admin-gradeable-updates.js:
```diff
@@ -325,7 +367,7 @@ function setRandomGraders(gradeable_id,p_values,successCallback,errorCallback) {
             }
             alert("error occured"+msg);
         }
-    });
```
Please remove whitespace

josharsh replied:
Corrected!

josharsh commented:
@bmcutler This needs your review :)

josharsh commented:
@emaicus  Yes! It looks so complex right now and I am already working on making it modular and concise in a different PR (which we also discussed last week) which also includes using different way to access students than using grading order. 
Will that work fine?

bmcutler commented:
@josharsh please resolve conflicts

josharsh commented:
@bmcutler Resolved!
